docs/src/references/agent/operation-workflow.md

```diff
@@ -0,0 +1,263 @@
+    A software component added by the user might need these *extra* fields the plugin is not aware of.
+- A workflow implementation must not react on *no-op* states nor terminal states.
+  - The transition from a *no-op* state must be handled either by thin-edge as direct transition
+    or overridden by the user with domain-specific checks and actions.
+  - The terminal states, a.k.a **successful** and **failed**, are owned by the process which created the **init** state (in practice, the mapper).
+    Only this process should clear the retained message state for an operation instance by sending an empty payload on command's topic.
+
+### Workflow Overriding
+
+Thin-edge provides a mechanism to override, extend and combine workflows.
+
+This mechanism is provided by the **tedge-agent** which gathers a set of user-defined workflows
+and combined them with the builtin workflows implemented by the agent itself.
+
+Each workflow is defined using a TOML file stored in `/etc/tedge/operations`. Each specifies:
+- the command name that it should trigger on
+  such as `firmware_update` or `restart`- the list of states
+- for each state:
+  - the state name as defined by the operation API
+  - the set of states which can be an outcome for this state actions
+  - possible extra instructions on how to process the command at this stage, e.g.
+    - run a script
+
+```toml title="file: firmware_update_example.toml"
+operation = "firmware_update"
+
+[init]
+  script = "/usr/bin/firmware_handler.sh plan"
+  next = ["scheduled", "failed"]
+
+[scheduled]
+  next = ["install"]
+
+[install]
+  script = "/usr/bin/firmware_handler.sh install ${.payload.url}"
+  next = ["reboot", "failed"]
+
+[reboot]
+  next = "verify"
+
+[verify]
+  script = "/usr/bin/firmware_handler.sh verify"
+  next = ["commit", "rollback"]
+
+[commit]
+  next = ["successful", "rollback_reboot"]
+
+[rollback]
+  script = "/usr/bin/firmware_handler.sh rollback"
+  next = ["rollback_reboot"]
+
+[rollback_reboot]
+  next = ["failed"]
+
+[successful]
+  next = []
+
+[failed]
+  next = []
+```
+
+Thin-edge combines all these workflows to determine what has to be done
+when a state message is published for a command on one topic matching the global topic filter for commands,
+i.e. `te/+/+/+/+/cmd/+/+`. Given an actual command state message, thin-edge
+- keeps only the workflows which target topic filter matches the topic of the message
+- then keeps only the workflows that define a behavior for the state of the message payload
+- and then uses priority rules to select a single workflow: the workflow ruling this command at that stage
+
+The priority rules give a higher priority to the workflow that are user-defined than to those pre-defined by thin-edge.
+If several user-defined workflows are matching a command state,
+then the alphabetic order of the workflow definition file names is used:
+`001-firmware_update_workflow.toml` being of higher priority than `002-firmware_update_workflow.toml`.
```

Comment:
This has not been implemented and has to be removed.

In the current settings, this is an error to provide two definitions for the same operation on a given device.

Reply:
Fixed with https://github.com/thin-edge/thin-edge.io/pull/2071/commits/ce7a46b588ffa78c9e2d23fbb9ee205cd614b959